File: backend/app/api/v1/cm.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_db_session, require_role
from app.models.user import User, UserRole
from app.schemas.cm import DisputeDetailRead, DisputeResolveRequest, DisputeResolveResponse, MatrixProposalDecideResponse, MatrixProposalDetailRead, PendingActionsResponse, PromotionDecideRequest, PromotionDecideResponse, PromotionDetailRead, SpecialistCardRead, SpecialistDetailRead
from app.services import cm_service

router = APIRouter()


@router.get("/team", response_model=list[SpecialistCardRead])
async def get_cm_team(
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> list[SpecialistCardRead]:
    return await cm_service.get_team_overview(current_user.id, db)


@router.get("/specialists/{specialist_id}", response_model=SpecialistDetailRead)
async def get_specialist_detail(
    specialist_id: UUID,
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> SpecialistDetailRead:
    return await cm_service.get_specialist_detail(current_user.id, specialist_id, db, page, per_page)


@router.get("/pending", response_model=PendingActionsResponse)
async def get_cm_pending(
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> PendingActionsResponse:
    return await cm_service.get_pending_actions(current_user.id, db)


@router.get("/disputes/{dispute_id}", response_model=DisputeDetailRead)
async def get_cm_dispute(
    dispute_id: UUID,
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> DisputeDetailRead:
    return await cm_service.get_dispute_detail(current_user.id, dispute_id, db)


@router.post("/disputes/{dispute_id}/actions/resolve", response_model=DisputeResolveResponse)
async def resolve_cm_dispute(
    dispute_id: UUID,
    body: DisputeResolveRequest,
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> DisputeResolveResponse:
    return await cm_service.resolve_dispute(current_user.id, dispute_id, body, db)


@router.get("/promotions/{notification_id}", response_model=PromotionDetailRead)
async def get_cm_promotion(
    notification_id: UUID,
    page: int = Query(1, ge=1),
    per_page: int = Query(10, ge=1, le=100),
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> PromotionDetailRead:
    return await cm_service.get_promotion_detail(current_user.id, notification_id, db, page, per_page)


@router.post("/promotions/{notification_id}/actions/approve", response_model=PromotionDecideResponse)
async def approve_cm_promotion(
    notification_id: UUID,
    body: PromotionDecideRequest,
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> PromotionDecideResponse:
    return await cm_service.approve_promotion(current_user.id, notification_id, body, db)


@router.post("/promotions/{notification_id}/actions/reject", response_model=PromotionDecideResponse)
async def reject_cm_promotion(
    notification_id: UUID,
    body: PromotionDecideRequest,
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> PromotionDecideResponse:
    return await cm_service.reject_promotion(current_user.id, notification_id, body, db)


@router.get("/matrix-proposals/{proposal_id}", response_model=MatrixProposalDetailRead)
async def get_cm_matrix_proposal(
    proposal_id: UUID,
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> MatrixProposalDetailRead:
    return await cm_service.get_matrix_proposal_detail(current_user.id, proposal_id, db)


@router.post("/matrix-proposals/{proposal_id}/actions/approve", response_model=MatrixProposalDecideResponse)
async def approve_cm_matrix_proposal(
    proposal_id: UUID,
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> MatrixProposalDecideResponse:
    return await cm_service.approve_matrix_proposal(current_user.id, proposal_id, db)


@router.post("/matrix-proposals/{proposal_id}/actions/reject", response_model=MatrixProposalDecideResponse)
async def reject_cm_matrix_proposal(
    proposal_id: UUID,
    db: AsyncSession = Depends(get_db_session),
    current_user: User = Depends(require_role(UserRole.CM)),
) -> MatrixProposalDecideResponse:
    return await cm_service.reject_matrix_proposal(current_user.id, proposal_id, db)

File: frontend/src/api/cm.ts
import type { DisputeDetailData, MatrixProposalDecideResponse, MatrixProposalDetail, PendingActionsData, PromotionDecideRequest, PromotionDecideResponse, PromotionDetailData, ResolveDisputeRequest, ResolveDisputeResponse, SpecialistCard, SpecialistDetail } from '../types/domain'
import { apiClient } from './client'

export async function getCmTeam(): Promise<SpecialistCard[]> {
  const response = await apiClient.get<SpecialistCard[]>('/cm/team')
  return response.data
}

export async function getCmSpecialistDetail(
  specialistId: string,
  page = 1,
  perPage = 20,
): Promise<SpecialistDetail> {
  const response = await apiClient.get<SpecialistDetail>(
    `/cm/specialists/${specialistId}`,
    { params: { page, per_page: perPage } },
  )
  return response.data
}

export async function getCmPending(): Promise<PendingActionsData> {
  const response = await apiClient.get<PendingActionsData>('/cm/pending')
  return response.data
}

export async function getCmDisputeDetail(disputeId: string): Promise<DisputeDetailData> {
  const response = await apiClient.get<DisputeDetailData>(`/cm/disputes/${disputeId}`)
  return response.data
}

export async function resolveCmDispute(
  disputeId: string,
  data: ResolveDisputeRequest,
): Promise<ResolveDisputeResponse> {
  const response = await apiClient.post<ResolveDisputeResponse>(
    `/cm/disputes/${disputeId}/actions/resolve`,
    {
      decision: data.decision,
      cm_note: data.cmNote ?? null,
      override_score: data.overrideScore ?? null,
    },
  )
  return response.data
}

export async function getCmPromotionDetail(notificationId: string): Promise<PromotionDetailData> {
  const response = await apiClient.get<PromotionDetailData>(`/cm/promotions/${notificationId}`)
  return response.data
}

export async function approveCmPromotion(
  notificationId: string,
  data: PromotionDecideRequest,
): Promise<PromotionDecideResponse> {
  const response = await apiClient.post<PromotionDecideResponse>(
    `/cm/promotions/${notificationId}/actions/approve`,
    { cm_note: data.cmNote ?? null },
  )
  return response.data
}

export async function rejectCmPromotion(
  notificationId: string,
  data: PromotionDecideRequest,
): Promise<PromotionDecideResponse> {
  const response = await apiClient.post<PromotionDecideResponse>(
    `/cm/promotions/${notificationId}/actions/reject`,
    { cm_note: data.cmNote ?? null },
  )
  return response.data
}

export async function getMatrixProposalDetail(proposalId: string): Promise<MatrixProposalDetail> {
  const response = await apiClient.get<MatrixProposalDetail>(`/cm/matrix-proposals/${proposalId}`)
  return response.data
}

export async function approveMatrixProposal(proposalId: string): Promise<MatrixProposalDecideResponse> {
  const response = await apiClient.post<MatrixProposalDecideResponse>(
    `/cm/matrix-proposals/${proposalId}/actions/approve`,
    {},
  )
  return response.data
}

export async function rejectMatrixProposal(proposalId: string): Promise<MatrixProposalDecideResponse> {
  const response = await apiClient.post<MatrixProposalDecideResponse>(
    `/cm/matrix-proposals/${proposalId}/actions/reject`,
    {},
  )
  return response.data
}
